### PR TITLE
[RDY] vim-patch: 7.4.2212, 7.4.2237, 7.4.2309, 7.4.2312

### PR DESCRIPTION
**Pull request by gregorias**
Problem:    Crash when autocommand moves to another tab. (Dominique Pelle)
Solution:   When navigating to another window halfway the :edit command go
            back to the right window.

https://github.com/vim/vim/commit/5a49789a9b1f6447aeafbbbdd5b235dd10c471d5

### CODE REVIEW

**Comment by gregorias** [comment]
I'm not sure whether I should and how to port the additional tests introduced in this patch. Could you help me out?

**Comment by KillTheMule** [comment]
The simplest thing would probably be to also port [patch 7.4.2227](https://github.com/vim/vim/commit/1381d791478ee77c8368b89a3d4954316bf839b5), which introduces `test_tabpage.vim`, so you can add your changes just like upstream. If you feel like doing a complete round, you could also port 2237 and 2309 with this batch (see the [history of `test_tabpage.vim`](https://github.com/vim/vim/commits/5a49789a9b1f6447aeafbbbdd5b235dd10c471d5/src/testdir/test_tabpage.vim)), but that's certainly not mandatory.

**Comment by Shougo** [comment]
I think 7.4.2312, 2313, 2314 is set changes.

**Comment by gregorias** [comment]
According to the [wiki](https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-Vim) I shouldn't port this test file, so I'd like to skip this one.

I found that I actually need a different patch #5505, before merging this one. This PR doesn't even build right now, I should have checked it.

**Comment by KillTheMule** [comment]
I'd say it'd be better both for reviewing as well as the git history if you included all those related patches in this PR here. This will also speed up the process somewhat. Feel free to force push to your feature branch freely :)

File: src/nvim/testdir/test_tabpage.vim [hunk 1]
```diff
@@ -186,4 +186,50 @@ function Test_tabpage_with_autocmd()
...
+  call assert_fails('99tab help', 'E16:')
+  call assert_fails('+99tab help', 'E16:')
+  call assert_fails('-99tab help', 'E16:')
+
+  delfunction s:check_tab
+  tabonly!
+  bw!
+endfunction
+
+func Test_tabnext_on_buf_unload()
```
Comment (justinmk) [inline src/nvim/testdir/test_tabpage.vim]: this was renamed to `Test_tabnext_on_buf_unload1()` in https://github.com/vim/vim/commit/5a49789a9b1f6447aeafbbbdd5b235dd10c471d5

and we are missing `Test_tabnext_on_buf_unload2()` from that patch.

Comment (justinmk) [inline src/nvim/testdir/test_tabpage.vim]: I'll fix this while merging.

**Comment by justinmk** [comment]
Verified (some tests). Merged.

@gregorias Thanks! Nice choices (such as FUNC_ attributes and shada lua test).

**Pull request merged**